R/Phenotyping.R

```diff
@@ -348,4 +351,84 @@ ExpandGeneList <- function(genes, verbose = TRUE) {
+  unique_gene_set_pairs <- unique.data.frame(gene_matches_df[,c("gene", "geneset_union")])
+  
+  return(unique_gene_set_pairs)
+}
+
+#' @title MakePhenotypingDotPlot
+#'
+#' @description Creates a DotPlot using custom gene sets and attempts to coarsely group gene sets by cell type. 
+#' @param yField The grouping variable used to calculate the average expression of genes and the y axis of the DotPlot.
+#' @param scaled A boolean defining whether to color dots by scaled expression or unscaled expression.
+#' @param gene_lists A vector of gene lists (defined by .RegisterGeneSet) to be queried and their genes be plotted.
+MakePhenotypingDotPlot <- function(seuratObj,
+                                   yField = 'ClusterNames_0.2',
+                                   scaled = T, 
+                                   gene_lists = c('Cytotoxicity', 'EffectorCytokines')
+){
+  if (!is.logical(scaled)){
+    warning("Please ensure scaled is either TRUE (to use per-gene scaled expression) or FALSE (for raw gene expression)")
+    stop()
+  }
+  #Parse gene_lists and coerce into a vector of genes to be plotted
+  meta_gene_vector <- unique(unlist(sapply(gene_lists, FUN = RIRA::GetGeneSet)))
+  
+  #Get initial plotting and expression data from Seurat's version of the DotPlot
+  plt <- Seurat::DotPlot(seuratObj, features =  meta_gene_vector, group.by = yField)
+  dotplot_df <- plt$data
+  
+  #Begin Phenotyping
+  dotplot_df$Phenotype <- 0 #pass 0 as an easy debugging test for un/underannotated gene sets
+  matching_genesets <- .FindGeneSetsContaining(dotplot_df$features.plot)
+  dotplot_df <- merge(dotplot_df, matching_genesets, by.x = "features.plot", by.y ='gene')
+  
+  #Parse Phenotype Field to determine which cell type the phenotype targets
+  dotplot_df$CellType <- "Unknown_CellType" #set default celltype as unknown
+  dotplot_df[grepl("MemoryAndNaive", dotplot_df$geneset_union), "CellType"] <- "Cell Type: T Cells"
+  dotplot_df[grepl("Myelocytes|Pro_Myelocytes", dotplot_df$geneset_union), "CellType"] <- "Cell Type: Neutrophil Precursors"
+  dotplot_df[grepl("TandNK_Activation|Cytotoxicity|EffectorCytokines|Exhaustion", dotplot_df$geneset_union), "CellType"] <- "Cell Type: T Cells / NK Cells"
+  
+  #Sort the dataframe for faceting
+  dotplot_df$CellType <- naturalsort::naturalfactor(dotplot_df$CellType)
+  dotplot_df <- dotplot_df |> arrange(CellType)
+  
+  #Determine if the data should be scaled or not, then alter the color scheme accordingly
+  if(scaled){
+    colorField <- sym("avg.exp.scaled")
+    colors <- c("blue", "white", "red")
+    colorLabel = 'Scaled Average Expression'
+  } else if(!scaled) {
+    colorField <- sym("avg.exp")
+    colors = c("navy", "gold", "orange", "red")
+    colorLabel = 'Average Expression'
+  }
+
+  P1 <- ggplot(dotplot_df, aes(x = features.plot, y = id, size = pct.exp, color = !!colorField)) + 
+    geom_point() + 
+    facet_wrap(~CellType+geneset_union, scales = "free_x") + 
```

Comment:
Additionally, I've been waffling on how to properly facet these. The GeneSet names are pretty long, so `facet_grid(...,scales = "free_x")` end up truncating the gene set. The alternative is sometimes rather empty panels, but everything is clearly legible. 

e.g.:
<img width="210" alt="Screen Shot 2022-07-25 at 11 20 04 AM" src="https://user-images.githubusercontent.com/7402210/180846911-63e8aa96-6045-4906-a1bf-884e9c41af42.png">
 
 
 I think supporting custom `gene_lists` and then just interactively changing the `gene_lists` argument is good enough, but I'm definitely open to other ideas.